File: backend/app/schemas/progress.py
from datetime import datetime

from pydantic import BaseModel, Field


class ProgressTodoItem(BaseModel):
    id: str
    text: str


class ProgressUpdateLog(BaseModel):
    id: str
    title: str
    description: str


class ProgressRepoLog(BaseModel):
    id: str
    repo_label: str
    repo_full_name: str
    updates: list[ProgressUpdateLog] = Field(default_factory=list)


class ProgressGenerationRepo(BaseModel):
    repo_label: str
    repo_full_name: str
    base_ref: str | None = None
    head_ref: str
    commit_count: int = 0


class ProgressGenerationMeta(BaseModel):
    generator: str = "codex"
    scope: str = "committed"
    grouping: str = "repo_updates"
    granularity: str = "summarized_from_pr_and_commit"
    generated_at: datetime | None = None
    repos: list[ProgressGenerationRepo] = Field(default_factory=list)


class ProgressDay(BaseModel):
    id: str
    date: str
    title: str
    repos: list[ProgressRepoLog] = Field(default_factory=list)


class ProgressBoardPayload(BaseModel):
    todo_items: list[ProgressTodoItem] = Field(default_factory=list)
    published_progress_days: list[ProgressDay] = Field(default_factory=list)


class ProgressDraftPayload(BaseModel):
    draft_progress_days: list[ProgressDay] = Field(default_factory=list)
    generation_meta: ProgressGenerationMeta | None = None


class ProgressTodoPayload(BaseModel):
    todo_items: list[ProgressTodoItem] = Field(default_factory=list)


class ProgressBoardResponse(ProgressBoardPayload):
    draft_progress_days: list[ProgressDay] | None = None
    draft_generation_meta: ProgressGenerationMeta | None = None
    published_generation_meta: ProgressGenerationMeta | None = None
    updated_at: datetime | None = None
    updated_by_user_id: int | None = None
    updated_by_name: str | None = None
    last_synced_at: datetime | None = None
    last_synced_by_user_id: int | None = None
    last_synced_by_name: str | None = None
    last_sync_status: str = "never"
    last_sync_error: str | None = None
    last_published_at: datetime | None = None
    last_published_by_user_id: int | None = None
    last_published_by_name: str | None = None

File: frontend/src/types/progress.ts
export interface ProgressTodoItem {
  id: string
  text: string
}

export interface ProgressUpdateLog {
  id: string
  title: string
  description: string
}

export interface ProgressRepoLog {
  id: string
  repo_label: 'FIT' | 'AkShare Project' | string
  repo_full_name: string
  updates: ProgressUpdateLog[]
}

export interface ProgressGenerationRepo {
  repo_label: 'FIT' | 'AkShare Project' | string
  repo_full_name: string
  base_ref: string | null
  head_ref: string
  commit_count: number
}

export interface ProgressGenerationMeta {
  generator: 'codex' | string
  scope: 'committed' | string
  grouping: 'repo_updates' | string
  granularity: 'summarized_from_pr_and_commit' | string
  generated_at: string | null
  repos: ProgressGenerationRepo[]
}

export interface ProgressDay {
  id: string
  date: string
  title: string
  repos: ProgressRepoLog[]
}

export interface ProgressBoardResponse {
  todo_items: ProgressTodoItem[]
  published_progress_days: ProgressDay[]
  published_generation_meta: ProgressGenerationMeta | null
  draft_progress_days?: ProgressDay[] | null
  draft_generation_meta?: ProgressGenerationMeta | null
  updated_at: string | null
  updated_by_user_id: number | null
  updated_by_name: string | null
  last_synced_at: string | null
  last_synced_by_user_id: number | null
  last_synced_by_name: string | null
  last_sync_status: 'never' | 'success' | 'failed' | string
  last_sync_error: string | null
  last_published_at: string | null
  last_published_by_user_id: number | null
  last_published_by_name: string | null
}

export interface ProgressBoardPayload {
  todo_items: ProgressTodoItem[]
  published_progress_days: ProgressDay[]
}

export interface ProgressTodoPayload {
  todo_items: ProgressTodoItem[]
}

export interface ProgressDraftPayload {
  draft_progress_days: ProgressDay[]
  generation_meta?: ProgressGenerationMeta | null
}
